[RFC] Fix system('string') quoting in Windows

equalsraf opened the pull request:
This does not change the current behaviour of `system([...])` it only affects the string variant `system('...')` so that it is compatible with Vim (Vim does not have `system([])`). 

This PR disables the libuv quoting when spawning the new process - libuv quotes according to `CommandLineToArgvW()` however cmd.exe does not follow this convention. A good article about the differences can be found here

https://blogs.msdn.microsoft.com/twistylittlepassagesallalike/2011/04/23/everyone-quotes-command-line-arguments-the-wrong-way/

This fixes issues when calling the shell (#6329) in particular it should restore compatibility with Vim in this regard. It does however make `system('...')` less usable with regards to other windows shells that use `CommandLineToArgvW()` but these difficulties are also seen in Vim.

TODO;
- [x] add note in system() docs about system(['cmd.exe', '/c', 'echo ""']) not producing the expected result

justinmk committed:
win: defaults: 'shellredir', 'shellxquote', 'shellxescape'

justinmk committed:
win: vim_strsave_shellescape: Handle 'shellslash'.

From Vim, misc2.c:vim_strsave_shellescape

Commit:
Windows: Disable process argument quoting when calling the shell

In Windows libuv quotes the arguments when spawning new processes. Add
an option to the Process struct to enable/disable quoting, by default it
is enabled.

This commit disables argument quoting when executing shell commands

- jobstart()/system() disable quoting if the command argument is a
  string
- Quoting in always enabled for rpcstart()
- Quoting is always disabled for os_call_shell()

Commit:
Return cmdshell=true if tv_to_argv() uses the shell

If cmd is a string, tv_to_argv() build a shell command using &shell.
Set bool ptr argument to true if the returned command uses the shell.

Commit:
Windows: Enable more system() functional tests

equalsraf commented:
I've reenabled most system() tests (except one, unrelated issue).

The failling jobstart() tests seem to be related to the use of powershell in those tests. I have not confirmed, but it would make some sense to assume that with the removal of libuv quoting arguments are no longer correctly escaped for a call to powershell e.g.

```vim
" as used in job_spec.lua
set shell=powershell shellquote=\" shellpipe=\| shellredir=>
set shellcmdflag=\ -ExecutionPolicy\ RemoteSigned\ -Command
let &shellxquote=' '
" this prints two lines instead of one
echo system('echo "one more"')
" but this prints a single line
echo system(shellescape('echo "one more"'))
```

I've executed the previous example in GVim and the result was identical to that of Neovim.

Notice the tests use the first form e.g. https://github.com/neovim/neovim/blob/master/test/functional/core/job_spec.lua#L389 adding a `shellescape()` fixes the issue (minus a trailling newline).

So at this point we need to have a hard look at what is the expected behaviour for `system('...')` in Windows, and put down the necessary functional tests to pin it down. I assume the idea is to keep compatibility with Vim.

Commit:
Windows: shellescape() jobstart test commands

With the change in process argument quoting for system('...') in
Windows, calling powershell via system() with whitespaces inside the
command produces unexpected results.

The jobstart tests use powershell in Windows and need to be changed
accordingly, otherwise the argument string is treated as multiple
arguments and the output is different from expected.

fixup

equalsraf commented:
Fixed the jobstart tests. Some lint issues remain.

It seems tests are hanging in travis. Could this be caused by shellescape()? Pushing a revert to check.

janlazo commented:
Why not add -NoProfile to shellcmdflag to speed up the powershell tests?

Commit:
tests: Execute powershell with -NoProfile

To speed up execution call powershell with -NoProfile argument.

equalsraf commented:
@janlazo thank for the tip.

Retagging WIP as I'm pretty sure I should not be using `shellescape()` like that. How exactly are we supposed to configure shell options for powershell anyway?

janlazo commented:
Why use a space for shellxquote for powershell? Its behaviour is closer to sh than cmd.

```viml
let &shellxquote=' '
set noshellslash
echo system('echo "one more"')
```
```
one
more
```
double-quote to escape commands/arguments
```viml
echo system(shellescape('echo "one more"'))
echo system('"echo \"one more\""') 
```
```
one more
one more
```
single-quote to output a literal string
```viml
set shellslash
echo system(shellescape('echo "one more"'))
```
```
echo one more
```

equalsraf commented:
> Why use a space for shellxquote for powershell? Its behaviour is closer to sh than cmd.

Good question, I dont know why it was set that way in the functional tests. I think `shellxquote=\"` would make more sense and would allow for `system('echo '.shellescape('"one more"'))`

justinmk commented:
> I dont know why it was set that way in the functional tests

I copied it from somewhere. Feel free to change it.

equalsraf commented:
> I copied it from somewhere. Feel free to change it.

Will change in a bit. Not that it makes much difference with regards to this PR.

equalsraf commented:
> Will change in a bit. Not that it makes much difference with regards to this PR.

Nevermind. It really does not help in writing the tests, it always results in different cases for win32.

erw7 commented on src/nvim/eval.c:
```diff
@@ -12791,13 +12791,26 @@ static void f_jobresize(typval_T *argvars, typval_T *rettv, FunPtr fptr)
     if (cmd) {
       *cmd = cmd_str;
     }
+    if (shellcmd) {
```
Would not it be better to be 
if (shellcmd && *p_sxq != NUL) {

erw7 commented on src/nvim/eval.c:
```diff
@@ -24033,6 +24050,7 @@ static inline TerminalJobData *common_job_init(char **argv,
   }
   Process *proc = (Process *)&data->proc;
   proc->argv = argv;
+  proc->quote_argv = quote_argv;
```
Would not it be better to be
proc->quote_argv = *p_sxq == NUL ? false : quote_argv;

equalsraf replied:
So the quoting would be disabled if `shellxquote` was not set. That is an interesting choice. Essentially using shellxquote to make this behaviour paremeterizable.

I cannot do it for the general case (in that line), because it would affect `jobstart([])`/`system([])`/etc and the goal of this PR is only to change the behaviour of the string variants `system('...')`/`jobstart('...')` - it makes sense for shellxquote to affect the later but not the former.

Could do it earlier as `!shellcmd  || p_sxq == NULL`

erw7 replied:
I understood the meaning of the flag in reverse. The code I intended was ```proc->quote_argv = *p_sxq == NUL ? true : quote_argv;```. This should not affect ```system([])/jobstart ([])```/etc.

How about making ```if (!proc->quote_argv && *p_sxq != NUL) {``` in libuv_process.c rather than ```!shellcmd || *p_sxq == NULL``` in multiple places?

equalsraf replied:
> How about making if (!proc->quote_argv && *p_sxq != NUL) { in libuv_process.c

Its kind of strange (from a logic perspective) to make a decision based on p_sxq so far from the reason behind it i.e. 6 months down the road if someone looks at the code in libuv_process.c and sees that what would they think?

equalsraf committed:
Add note to docs about jobstart(['cmd.exe'])

In Windows calling system([...]) follows the convention from the
CommandLineToArgvW API and used by most windows programs. However
cmd.exe does not, and its an important enough case to deserve a quick
note in the docs.

janlazo commented:
Why not disable all quoting for powershell? It can handle paths with spaces and backslashes unlike cmd which requires double-quotes.
```viml
set noshellslash
set shellquote= shellxquote= shellxescape=
```

justinmk commented:
> In powershell, why not disable quoting for paths without special characters such as parentheses? It can handle paths with spaces and backslashes unlike cmd which requires double-quotes.

We should be explicit if possible, to avoid edge cases. How can powershell know if two paths are intended to be a single path with a space or not?

janlazo commented:
@justinmk via backslashes. If using forward slashes, single/double quotes are needed.

equalsraf committed:
Add functional tests for system() in windows

justinmk commented:
Closing in favor of #6387.
